File: codes/java/438.找到字符串中所有字母异位词.java

class Solution {
	public List<Integer> findAnagrams(String s, String t) {
		Map<Character, Integer> need = new HashMap<>();
		Map<Character, Integer> window = new HashMap<>();
		for (char c : t.toCharArray())
			need.put(c, need.getOrDefault(c, 0) + 1);

		int left = 0, right = 0;
		int valid = 0;
		List<Integer> res = new ArrayList<>();
		while (right < s.length()) {
			char c = s.charAt(right);
			right++;
			// 进行窗口内数据的一系列更新
			if (need.containsKey(c)) {
				window.put(c, window.getOrDefault(c, 0) + 1);
				if (window.get(c).equals(need.get(c)))
					valid++;
			}
			// 判断左侧窗口是否要收缩
			while (right - left >= t.length()) {
				// 当窗口符合条件时，把起始索引加入 res
				if (valid == need.size())
					res.add(left);
				char d = s.charAt(left);
				left++;
				// 进行窗口内数据的一系列更新
				if (need.containsKey(d)) {
					if (window.get(d).equals(need.get(d)))
						valid--;
					window.put(d, window.get(d) - 1);
				}
			}
		}
		return res;
	}
}

File: codes/python/438.找到字符串中所有字母异位词.py

class Solution:
    def findAnagrams(self, s: str, t: str) -> List[int]:
        need = {}
        window = {}
        for c in t:
            if c in need:
                need[c] += 1
            else:
                need[c] = 1

        left = 0
        right = 0
        valid = 0
        res = []
        
        while right < len(s):
            c = s[right]
            right += 1
            # 进行窗口内数据的一系列更新
            if c in need:
                if c in window:
                    window[c] += 1
                else:
                    window[c] = 1
                if window[c] == need[c]:
                    valid += 1
            # 判断左侧窗口是否要收缩
            while right - left >= len(t):
                # 当窗口符合条件时，把起始索引加入 res
                if valid == len(need):
                    res.append(left)
                d = s[left]
                left += 1
                # 进行窗口内数据的一系列更新
                if d in need:
                    if window[d] == need[d]:
                        valid -= 1
                    window[d] -= 1

        return res